sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala

```diff
@@ -246,85 +260,238 @@ object SpecifiedWindowFrame {
   }
 }
 
+case class UnresolvedWindowExpression(
+    child: Expression,
+    windowSpec: WindowSpecReference) extends UnaryExpression with Unevaluable {
+
+  override def dataType: DataType = throw new UnresolvedException(this, "dataType")
+  override def foldable: Boolean = throw new UnresolvedException(this, "foldable")
+  override def nullable: Boolean = throw new UnresolvedException(this, "nullable")
+  override lazy val resolved = false
+}
+
+case class WindowExpression(
+    windowFunction: Expression,
+    windowSpec: WindowSpecDefinition) extends Expression with Unevaluable {
+
+  override def children: Seq[Expression] = windowFunction :: windowSpec :: Nil
+
+  override def dataType: DataType = windowFunction.dataType
+  override def foldable: Boolean = windowFunction.foldable
+  override def nullable: Boolean = windowFunction.nullable
+
+  override def toString: String = s"$windowFunction $windowSpec"
+}
+
 /**
- * Every window function needs to maintain a output buffer for its output.
- * It should expect that for a n-row window frame, it will be called n times
- * to retrieve value corresponding with these n rows.
+ * A window function is a function that can only be evaluated in the context of a window operator.
  */
 trait WindowFunction extends Expression {
-  def init(): Unit
+  /** Frame in which the window operator must be executed. */
+  def frame: WindowFrame = UnspecifiedFrame
+}
 
-  def reset(): Unit
+/**
+ * An offset window function is a window function that returns the value of the input column offset
+ * by a number of rows within the partition. For instance: an OffsetWindowfunction for value x with
+ * offset -2, will get the value of x 2 rows back in the partition.
+ */
+abstract class OffsetWindowFunction
+  extends Expression with WindowFunction with Unevaluable with ImplicitCastInputTypes {
+  val input: Expression
+  val default: Expression
+  val offset: Expression
+  val offsetSign: Int
+
+  override def children: Seq[Expression] = Seq(input, offset, default)
 
-  def prepareInputParameters(input: InternalRow): AnyRef
+  override def foldable: Boolean = input.foldable && (default == null || default.foldable)
```

Comment:
It can be foldable, for instance: ```LEAD(1, 2) OVER (PARTITION BY ... ORDER BY ...)``` should always return ```2```.

However there is currently no Optimizer rule to eliminate this. Should I add one?